In [4]:
# importing the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [5]:
data=pd.read_csv('50_Startups.csv')
data.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [9]:
# To check if there are any blank or null values in the dataset
data.isna().sum()

R&D Spend          0
Administration     0
Marketing Spend    0
State              0
Profit             0
dtype: int64

In [12]:
# For a specific column
data.dtypes

R&D Spend          float64
Administration     float64
Marketing Spend    float64
State               object
Profit             float64
dtype: object

In [14]:
# Assign independent variables (features) and dependent variable (target)
X = data[["Marketing Spend"]]
y = data["Profit"]

In [15]:
# dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [16]:
# Initialize a Linear Regression model using the LinearRegression class from sklearn
lr = LinearRegression()

In [17]:
# Train the Linear Regression model using the training data (X_train as features and y_train as target values)
lr.fit(X_train.values,y_train)

LinearRegression()

In [18]:
# Training the Linear Regression model to predict target values for the test dataset (X_test)
y_pred = lr.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


In [19]:
# To Calculate the R-squared score (coefficient) to evaluate the model's performance
r2_score( y_test, y_pred)

-0.11205578997827947

In [22]:
#  rounded to 3 decimal places
round(r2_score( y_test, y_pred),3)

-0.112

In [27]:
print(round(lr.intercept_,3))

48923.313


In [28]:
# Predict the target value for the first instance (row) of the dataset X using the trained Linear Regression model
lr.predict(X[0:1])


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([182462.19522503])

In [29]:
# Check the type of the trained Linear Regression model (lr)
type(lr)

sklearn.linear_model._base.LinearRegression

In [30]:
# Import the pickle module to serialize and deserialize Python objects, such as saving and loading machine learning models
import pickle

In [31]:
# Save the trained Linear Regression model (lr) to a file ('model.pkl') using pickle
# 'wb' mode opens the file in write-binary mode to store the serialized model
with open('model.pkl','wb') as f:
    pickle.dump(lr,f)

In [32]:
# Load the saved Linear Regression model ('model.pkl') from the file into the variable clf2 using pickle
# 'rb' mode opens the file in read-binary mode to deserialize the model
with open('model.pkl', 'rb') as f:
    clf2 = pickle.load(f)

In [33]:
X[0:5]

Marketing Spend
0        471784.10
1        443898.53
2        407934.54
3        383199.62
4        366168.42

In [34]:
clf2.predict(X[0:5])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([182462.19522503, 174569.16142223, 164389.52411328, 157388.28441585,
       152567.58907765])

In [35]:
# Create a DataFrame to compare the actual target values (y_test) with the predicted values (y_pred)
# Display the first few rows of the DataFrame for inspection
predictions = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})
print(predictions.head())

       Actual      Predicted
13  134307.35  120440.331743
39   81005.76   98457.010516
30   99937.59   74718.086315
45   64926.08   49462.221939
17  125370.37  128906.206651


Analysis of Simple Linear Regression:

The low R-squared value (-0.112) indicates the model is not capturing the
relationship between marketing spend and profit effectively.
Predictions are made, but the differences between actual and predicted values shows that the model may need improvement,by including more features (independent variables).
The intercept suggests a baseline profit when marketing spend is 0.
The model is functional but doesn't have high predictive accuracy.
The R2 score is -0.112 indicates, the model doesn't fit the data accurately.
